# Home assignment 2

You should work on the assignement in groups of 2 participants. 

Upload your solution as a jupyter notebook to L2P by 26th of June 23:59h. (The deadline is strict)

Do not forget to specify the names of all contributing students in the jupyter notebook.

You should add comments to your code where necessary and print the relevant results.

In [ ]:
Xiaoqi Ma (383420)
Anna Wegmann (330509)

# 1. Dynamic PageRank
Consider a random walk setting where the transistion matrix changes over time. At any point of time the probability of a random surfer to jump to a linked page is proportional to the number of previous visits. To start with all the pages are equally likely to be chosen but as the walk continues and the nodes are visited the transition probability changes as proportional to number of previous visits. For example let a page 'a' is linked to pages 'b', 'c' and 'd'. The random surfer currently resides at 'a' and the pages 'b', 'c' and 'd' have already been visited 5, 3 and 2 times respectively. The transition probability would be 0.5, 0.3 and 0.2 respectively. As a new node is viited the probabilities change. The random surfer continues to surf with probability 0.8. Generate 100 random walks and rank the nodes based on the frequency of visit. The random walk should be performed on a drected Erdos-Renyi graph with number of nodes n=200 and probability of edge creation p = 0.4. 

Hint: Use networkx library for generating graph.
    

In [1]:
import networkx as nx
import random
import numpy as np

# Create a directed Erdos-Renyi graph with 200 nodes, prob of edge creation is p=0.4
G = nx.erdos_renyi_graph(200, 0.4, directed=True)

In [2]:
# For each random walk, execute randon_surfer function with probability 0.8 to continue
# executesd one step for a random surfer, returns the current node and the visit frequency
def random_surfer(curr_node, visit_freq):
    node_prob = {} # dict to store probability for each node to be visited
    freq_sum = 0 # the sum of total visit frequency of other nodes that connect to curr_node
    nbrs = list(G.neighbors(curr_node)) # List all connected neighbors of current node
    if len(nbrs) > 0:
        for nb in nbrs: # for all neighbors
            node_prob[nb] = visit_freq[nb]
            freq_sum += visit_freq[nb]
        for k,v in node_prob.items():
            # To get each node probability to be visited
            node_prob[k] = v / freq_sum 
        # randomly choose the next node according to the probability
        # random.choice(a,size=None,replace=True,p=None) Generates a random sample from a given 1-D array
        next_node = np.random.choice(nbrs, 1, node_prob)[0] 
        visit_freq[next_node] += 1  
        #print('next node: ', next_node)
    return (next_node, visit_freq)

In [3]:
# Set stop probability because the surfer continues to surf with prob 0.8
p_stop = 0.2
visit_freq = {}
nodes = list(G.nodes())
# Initialize with all nodes as 'visited once' since it is devided by sum of freq in random_surfer
for n in nodes:
    visit_freq[n] = 1

walks = 100
for i in range(walks):
    # randomly choose a start node
    cur_node = random.choice(nodes)
    while random.uniform(0, 1) >= p_stop:
        # Execute random_surfer function
        #print('random surfer: ',i, ' going to node ', cur_node)
        (cur_node, visit_freq) = random_surfer(cur_node, visit_freq)

In [7]:
# Rank the nodes based on the visit frequency
counter = 0
for n,v in sorted(visit_freq.items(),key = lambda x:x[1], reverse = True): 
    print('node: ', n, ' visit frequency: ', v-1) # actual frequency is visit_freq-1
    counter += 1
    # Only output the 5 most frequently visited nodes
    if(counter == 5):
        break

node:  80  visit frequency:  8
node:  92  visit frequency:  8
node:  47  visit frequency:  6
node:  68  visit frequency:  6
node:  17  visit frequency:  5


In [6]:
# Output the sorted visit frequency dictionary
for m,n in sorted(visit_freq.items(),key = lambda x:x[1], reverse = True): 
    print('node {0}: {1} '.format(m, n-1))

node 80: 8 
node 92: 8 
node 47: 6 
node 68: 6 
node 17: 5 
node 27: 5 
node 46: 5 
node 57: 5 
node 66: 5 
node 75: 5 
node 78: 5 
node 117: 5 
node 138: 5 
node 155: 5 
node 160: 5 
node 173: 5 
node 3: 4 
node 5: 4 
node 33: 4 
node 44: 4 
node 49: 4 
node 53: 4 
node 56: 4 
node 63: 4 
node 69: 4 
node 81: 4 
node 84: 4 
node 89: 4 
node 105: 4 
node 113: 4 
node 135: 4 
node 143: 4 
node 154: 4 
node 190: 4 
node 192: 4 
node 6: 3 
node 11: 3 
node 12: 3 
node 15: 3 
node 20: 3 
node 24: 3 
node 32: 3 
node 40: 3 
node 41: 3 
node 51: 3 
node 61: 3 
node 70: 3 
node 71: 3 
node 76: 3 
node 77: 3 
node 86: 3 
node 87: 3 
node 121: 3 
node 123: 3 
node 127: 3 
node 128: 3 
node 132: 3 
node 133: 3 
node 137: 3 
node 147: 3 
node 157: 3 
node 158: 3 
node 163: 3 
node 179: 3 
node 180: 3 
node 185: 3 
node 186: 3 
node 191: 3 
node 4: 2 
node 7: 2 
node 8: 2 
node 9: 2 
node 13: 2 
node 16: 2 
node 18: 2 
node 21: 2 
node 22: 2 
node 31: 2 
node 34: 2 
node 37: 2 
node 42: 2 
node 54

# 2. Recommendation
a. Compare the recommendation algorithms (SVD, NMF, Baseline, k-NN and Random) available in surprise package on movielens dataset in terms of RMSE and MAE.

b. Consider the movielens dataset and divide it into (i) training set with 50% of the data (train the algorithms on this part) and (ii) 25% validation set and (iii) test set with the rest. Estimate the ratings of the test set using the algorithms (same as in a) provided by the package on the training set. Your final rating should be weighted average of the ratings predicted by the algorithms. The weights should be learnt on the validation set. Performance should be measured in terms of RMSE.

Hint: Use grid search/step-wise update like SGD for learning the weights

In [8]:
# -- 2a: Compare recommendation algorithms on movielens dataset (here: use movielens-100k)
# RMSE: Root Mean Square error (higher weight to large errors)
# MAE: Mean absolute error
# oriented at example in surprise package: http://surpriselib.com/
# algorithms: https://surprise.readthedocs.io/en/stable/prediction_algorithms_package.html?highlight=k-NN

# load Data
from surprise import SVD
from surprise import Dataset
from surprise.model_selection import cross_validate
data = Dataset.load_builtin('ml-100k')

In [9]:
algo = SVD()
resultSVD = cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9320  0.9330  0.9440  0.9355  0.9295  0.9348  0.0050  
MAE (testset)     0.7335  0.7362  0.7449  0.7368  0.7339  0.7370  0.0041  
Fit time          4.79    4.90    4.85    5.16    4.83    4.91    0.13    
Test time         2.47    0.14    0.16    0.19    0.19    0.63    0.92    


SVD on movielens-100k in different runs: RMSE ~ 0.937, MAE ~ 0.74

In [10]:
from surprise import BaselineOnly
algo = BaselineOnly()
resultBaseline = cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Evaluating RMSE, MAE of algorithm BaselineOnly on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9468  0.9456  0.9382  0.9476  0.9410  0.9438  0.0036  
MAE (testset)     0.7513  0.7486  0.7444  0.7499  0.7472  0.7483  0.0024  
Fit time          0.22    0.24    0.25    0.24    0.25    0.24    0.01    
Test time         0.16    0.11    0.15    0.15    0.16    0.15    0.02    


Baseline on movielens-100k in different runs: RMSE ~ 0.94, MAE ~ 0.75

In [11]:
from surprise import NMF
algo = NMF()
resultNMF = cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9631  0.9601  0.9614  0.9609  0.9672  0.9625  0.0025  
MAE (testset)     0.7548  0.7574  0.7556  0.7560  0.7586  0.7565  0.0013  
Fit time          5.21    5.19    5.56    5.38    5.19    5.30    0.15    
Test time         0.14    0.12    0.12    0.17    0.17    0.14    0.02    


NMF on movielens-100k in different runs: RMSE ~ 0.96, MAE ~ 0.76

In [12]:
from surprise import KNNBasic
algo = KNNBasic()
resultKNN = cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE, MAE of algorithm KNNBasic on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9851  0.9830  0.9764  0.9772  0.9747  0.9793  0.0040  
MAE (testset)     0.7789  0.7762  0.7716  0.7705  0.7707  0.7736  0.0034  
Fit time          0.41    0.44    0.47    0.44    0.44    0.44    0.02    
Test time         3.51    3.30    3.39    3.32    3.40    3.38    0.08    


k-NN on movielens-100k in different runs: RMSE ~ 0.98, MAE ~ 0.77

In [13]:
from surprise import NormalPredictor
algo = NormalPredictor()
resultRandom = cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm NormalPredictor on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.5214  1.5200  1.5243  1.5069  1.5194  1.5184  0.0060  
MAE (testset)     1.2219  1.2177  1.2227  1.2119  1.2163  1.2181  0.0040  
Fit time          0.11    0.13    0.15    0.14    0.15    0.13    0.02    
Test time         0.13    0.14    0.18    0.19    0.18    0.16    0.03    


Random on movielens-100k in different runs: RMSE ~ 1.52, MAE ~ 1.22

Since the lower the RMSE/MAE values are the less errors there are in the algorithms: algorithms are already sorted from best to worst (SVD, Baseline, NMF, k-NN and Random) 

In [14]:
# -- 2b (data: user,movie,rating (1-5) )
# Divide Dataset into: 50% training set, 25% validation set, 25% test set??
from surprise.model_selection import train_test_split
trainset, valTestset = train_test_split(data, train_size=0.5)
validationset = valTestset[0:int(len(valTestset)/2)]
testset = valTestset[int(len(valTestset)/2):len(valTestset)]

In [15]:
# train algorithms on training set
from surprise import NMF, KNNBasic, SVD, NormalPredictor, BaselineOnly
svdAlgo = SVD()
nmfAlgo = NMF()
randomAlgo = NormalPredictor()
kNNAlgo = KNNBasic()
baselineAlgo = BaselineOnly()

svdAlgo.fit(trainset)
nmfAlgo.fit(trainset)
randomAlgo.fit(trainset)
kNNAlgo.fit(trainset)
baselineAlgo.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.
Estimating biases using als...


In [16]:
# calculate ratings on the validationset (here: we output the accuracy on this as well)
from surprise import accuracy

validationSVD = svdAlgo.test(validationset)
svdAccuracy = accuracy.rmse(validationSVD)

validationNMF = nmfAlgo.test(validationset)
nmfAccuracy = accuracy.rmse(validationNMF)

validationRandom = randomAlgo.test(validationset)
randomAccuracy = accuracy.rmse(validationRandom)

validationKNN = kNNAlgo.test(validationset)
kNNAccuracy = accuracy.rmse(validationKNN)

validationBaseline = baselineAlgo.test(validationset)
baselineAccuracy = accuracy.rmse(validationBaseline)

RMSE: 0.9570
RMSE: 0.9860
RMSE: 1.5075
RMSE: 0.9990
RMSE: 0.9495


In [18]:
# create a Matrix of the estimated ratings of the validationset - columns are the different algos
import numpy as np
svdEst = [x.est for x in validationSVD]
nmfEst = [x.est for x in validationNMF]
randomEst = [x.est for x in validationRandom]
kNNEst = [x.est for x in validationKNN]
baselineEst = [x.est for x in validationBaseline]

correctRating = np.transpose([x.r_ui for x in validationSVD])
# estMatrix includes estimated ratings of one algorithm as a column
estMatrix = np.column_stack((svdEst,nmfEst,randomEst,kNNEst,baselineEst)) # Stack 1-D arrays as columns into a 2-D array.

In [19]:
# Try to find minimum for: w_1*svdEst + w_2*nmfEst + w_3*randomEst + w_4*kNNEst + w_5*baselineEst - actualRating
# OR: estMatrix*w - actualRating 
# use SGDRegressor for learing of the weights (continous valus are permitted)
from sklearn.linear_model import SGDRegressor
clf = SGDRegressor(loss="squared_loss", penalty="l2", average=True)
clf.fit(estMatrix, correctRating)
w = clf.coef_
print('estimated weights: ', w)

estimated weights:  [ 0.30813112  0.20571974 -0.00292919  0.15563319  0.35188168]


/home/anna/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


In [20]:
# For Comparison purposes print the RMSE on the VALIDATION SET
print('RMSE on validation set for hybrid algorithm: ', np.sqrt(sum((estMatrix.dot(w)-correctRating)**2)/len(correctRating)))
print('This is only  marginally better than the RMSE value for the best algorithm on the validation set.')

RMSE on validation set for hybrid algorithm:  0.946023515053171
This is only  marginally better than the RMSE value for the best algorithm on the validation set.


In [21]:
# Test the hybrid algorithm on the testset:
# Use algorithms on the test set and add them up with the calculated weights
# We print the RMSE for the individual algorithms in the process
testPredictSVD = svdAlgo.test(testset)
svdTestAccuracy = accuracy.rmse(testPredictSVD)
testPredictNMF = nmfAlgo.test(testset)
nmfTestAccuracy = accuracy.rmse(testPredictNMF)
testPredictRandom = randomAlgo.test(testset)
randomTestAccuracy = accuracy.rmse(testPredictRandom)
testPredictkNN = kNNAlgo.test(testset)
kNNTestAccuracy = accuracy.rmse(testPredictkNN)
testPredictBaseline = baselineAlgo.test(testset)
baselineTestAccuracy = accuracy.rmse(testPredictBaseline)

svdTestEst = [x.est for x in testPredictSVD]
nmfTestEst = [x.est for x in testPredictNMF]
randomTestEst = [x.est for x in testPredictRandom]
kNNTestEst = [x.est for x in testPredictkNN]
baselineTestEst = [x.est for x in testPredictBaseline ]
estMatrix = np.column_stack((svdTestEst,nmfTestEst,randomTestEst,kNNTestEst,baselineTestEst))

correctTestRating = [x.r_ui for x in testPredictBaseline ]
hybridTestEst = estMatrix.dot(w)

RMSE: 0.9585
RMSE: 0.9996
RMSE: 1.5259
RMSE: 1.0102
RMSE: 0.9587


In [22]:
# calculate RMSE
print('RMSE on test set for hybrid algorith:', np.sqrt(np.sum((hybridTestEst-correctTestRating)**2)/len(correctTestRating)))
print('This - again - is only a little bit better than the RMSE value for the best algorithm on the test set.')

RMSE on test set for hybrid algorith: 0.9541404468456874
This - again - is only a little bit better than the RMSE value for the best algorithm on the test set.


# 3. Hidden Markov model
Consider the HMM package https://hmmlearn.readthedocs.io/en/latest/

a. Generate sequences with multinomial HMM (2 symbols and 4 hidden states) and given parameters. Start probability - {0.4,0.2,0.1,0.3}, Transition matrix - {{0.2,0.3,0.1,0.4},{0.3,0.3,0.2,0.2},{0.4,0.2,0.3,0.1},{0.2,0.3,0.1,0.4}}, Emission probability - {{0.2,0.8},{0.1,0.9},{0.5,0.5},{0.6,0.4}}.


b. Consider a sequence - {1 0 0 0 1 1 1 1 0 1 0 1 0 1 0 1 1 1 0 0 0 1 1 0 1 0 0 1 1 0 1}. Fit a multinomial HMM considering 4 states and obtain hidden state which is most likely to have generated the symbol


In [47]:
# -- 3a
# scikit-learn HMM module:
#     used http://scikit-learn.sourceforge.net/stable/modules/hmm.html for orentation

from hmmlearn import hmm
import numpy as np

startprobability = np.array([0.4, 0.2, 0.1, 0.3])
transmatrix = np.array([[0.2, 0.3, 0.1, 0.4], [0.3, 0.3, 0.2, 0.2], [0.4, 0.2, 0.3, 0.1],[0.2, 0.3, 0.1, 0.4]])
emissionprobability = np.array([[0.2,0.8],[0.1,0.9],[0.5,0.5],[0.6,0.4]])

model = hmm.MultinomialHMM(n_components=4)
model.n_symbols = 2
model.startprob_ = startprobability
model.transmat_ = transmatrix
model.emissionprob_ = emissionprobability

X,Z = model.sample(100)
print('sampled state sequence: ',Z)
print('with feature sequence: ', X.transpose()[0]) # transposed for readbility

sampled state sequence:  [0 2 0 2 3 0 3 0 3 1 0 0 3 1 0 2 1 2 0 1 2 0 0 0 3 0 1 1 0 1 3 0 3 3 1 0 3
 0 3 2 2 0 0 1 1 2 3 3 3 1 1 1 2 0 1 0 1 1 1 1 2 0 3 1 0 1 0 3 1 1 0 3 1 0
 1 2 1 0 0 3 3 1 0 3 1 1 3 3 1 1 1 3 2 3 3 1 0 2 0 1]
with feature sequence:  [1 1 1 0 0 0 1 1 0 1 1 0 1 1 1 0 1 0 1 1 1 1 1 1 0 1 0 1 0 1 0 1 0 0 1 1 1
 0 0 1 1 1 1 1 1 0 0 0 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 0 1 1 1 0 1 0 0 1 1 1 0 1 0 1 1 0 0 1 0]


In [51]:
# -- 3b
# Fit a HMM on a given sequence and obtain hidden state, which generated the symbol - {1 0 0 0 1 1 1 1 0 1 0 1 0 1 0 1 1 1 0 0 0 1 1 0 1 0 0 1 1 0 1}

model2 = hmm.MultinomialHMM(n_components=4)
#model2.n_symbols = 2 should be assumed
X = np.array([[1], [0], [0], [0], [1], [1], [1], [1], [0], [1], [0], [1], [0], [1], [0], [1], [1], [1], [0], [0], [0], [1], [1], [0], [1], [0], [0], [1], [1], [0], [1]])
print('number of samples: ', X.shape[0], ' number of features: ', X.shape[1])
model2.fit(X) 
Z2 = model2.predict(X)
print('predicted state sequence: ', Z2)

number of samples:  31  number of features:  1
predicted state sequence:  [0 2 1 2 0 3 1 0 2 0 2 0 2 0 2 0 3 0 2 1 2 1 0 2 0 2 3 1 0 2 0]


# 4. PrefixSpan 
Implement the Prefix algorithm for Sequential Pattern Mining.

In [52]:
import copy

In [53]:
# with influences from: http://sequenceanalysis.github.io/
# but tried to implement independently (since I looked at the code from github first, the structure is pretty similar...)

# example input of a sequence database
# list of sequences (sequences can contain another list of items that occur at the same time)
db = [
    [["a", "b"], ["c"], ["d"], ["e"]], # here a and b occur at the same time at time step 0
    [["b"], ["b"], ["b", "d"], ["e"]],
    [["c"], ["b"], ["c"], ["c"], ["a"]],
    [["b"], ["b"], ["b", "c", "c"]],
]

In [54]:
"""
    write sequence pattern as a string
    called to save support count in a dict
"""
def toStr(seqPattern):
    result = "["
    #print(seqPattern)
    for j in range(len(seqPattern)):
        result = result + "["
        for i in range(len(seqPattern[j])):
            result = result + seqPattern[j][i]
            if i!=len(seqPattern[j])-1:
                result = result + ","
        result = result + "]"
        if j!=len(seqPattern)-1:
            result = result + ","
    result = result + "]"
    return result

In [55]:
toStr([["a","b"],["c"]])

'[[a,b],[c]]'

In [75]:
""" project whole database to prefix

    prefix: is one timestep of a sequence
        here: [[]]-notation, but only ever called with one time step!
        I implemented this before I knew this is the case -> does more than it has to, and also more complicated...
        what it does more: projects to a prefix, which can consist of several time steps!
    
    if nextTimestep is true: jump over first timestep since prefix is supposed to happen in a new timestep
    if nextTimestep is false: can happen at the same timestep
    result: projected db: does not replace prefix with '_' as in the algorithm 
"""
def projectDatabase(database, prefix, nextTimestep=False):
    projectedDB = []
    for sequence in database:
        if (nextTimestep): # prefix supposed to happen in the next time step
            sequence = sequence[1:]
        
        # ********** project sequence ************* #
        sequenceProjection = []

        for startLeft in range(len(sequence)): # shift starting point of prefix
            if (startLeft+len(prefix)>len(sequence)): break # break if prefix does not fit after startLeft
            sequenceProjection = copy.deepcopy(sequence)[startLeft:] # copy sequence from startLeft to projection
            
            if not all(x in sequence[startLeft] for x in prefix[0]):
                sequenceProjection = []
                continue # if first time step of prefix is not contained in sequence that starts at startLeft, increase startLeft

            j = startLeft # j is current position that is checked against prefix time step i
            for i in range(len(prefix)-1): # this is acutally never called
                while(j<len(sequence)):
                    if not all(x in sequence[j+i+1] for x in prefix[i+1]): # check if prefix is contained in sequence
                        j+=1 # look for prefix timestep in next sequence timestep
                    else:
                        break
                if j==len(sequence):
                    sequenceProjection = []
                    break
                        
            if len(sequenceProjection)>0:
                break
        
        
        if len(sequenceProjection)>0:
            projectedDB.append(sequenceProjection)
            
    return projectedDB

In [90]:
validateProjectDB = [ [["a"], ["b", "c"], ["a", "c"], ["c"]],[['a']] ]

In [91]:
projectDatabase(validateProjectDB, ["a"], True)

[[['a', 'c'], ['c']]]

In [92]:
projectDatabase(validateProjectDB,[["a"],["b"]])

[[['a'], ['b', 'c'], ['a', 'c'], ['c']]]

In [93]:
projectDatabase([ [["b"], ["a"], ["c"], ["a"]] ], [["a"],["a"]])

[[['a'], ['c'], ['a']]]

In [94]:
testdb = projectDatabase(db,[['a']])
testdb

[[['a', 'b'], ['c'], ['d'], ['e']], [['a']]]

In [110]:
'''
    get all frequent Items from a database (chars), and their support
    newEvent: ignores the first timestep
    prefix: only consider one timestep!! (only relevant if countInNewEvent is false, i.e. second part of recursion)
'''

from collections import defaultdict #https://www.accelebrate.com/blog/using-defaultdict-python/
# support: number of sequences that have the items as a subsequence! (does not count occurences within sequence)
def getFrequentItems(database, min_sup, prefix=[], countInNewEvent=False):
    supports = defaultdict(int) # initiates every key with 0 --> no key errors
    # set countWithinTimestep, i.e. if it is supposed to happen simultaneously with prefix
    countWithinTimestep = False
    if (len(prefix)>0 and not countInNewEvent):
        countWithinTimestep = True
    #print('countWithinTimestep: ', countWithinTimestep)
    
    for sequence in database:
        if countInNewEvent: # item is supposed to happen in a new time step -> jump over first timestep
            sequence = sequence[1:]

        uniqueItems = set() # set saves items that occur in a sequence
        for timestep in sequence:
            if (countWithinTimestep):
                if all(x in timestep for x in prefix):
                    #print('prefix: ',prefix, ' and timestep: ', timestep)
                    #print('prefix in timestep: ', all(x in timestep for x in prefix))
                    for item in timestep:
                        if (not (item in prefix)):
                            uniqueItems.add(item)
            else:
                for item in timestep:
                    uniqueItems.add(item)
    
        #print('unique Item set:', uniqueItems, ' in timestep: ', timestep)
        for item in uniqueItems: #max support for every item: number of sequences
            supports[item] +=1
            
    result = defaultdict(int)
    for it, support in supports.items():
        if (support>=min_sup):
            result[it] = support
    return sorted(result.items()) # make sure the '>'-cond in prefixSpan works

In [111]:
getFrequentItems(db, 3)

[('b', 4), ('c', 3)]

In [112]:
getFrequentItems(testdb, 0, ['a'], True)

[('c', 1), ('d', 1), ('e', 1)]

In [113]:
""" high level algorithm call
    input: sequence database, minimum support
    output: dict of sequential patterns : count
"""
def prefixSpan(sequence_DB,min_sup):
    result = {}
    
    frequentItems = getFrequentItems(sequence_DB, min_sup)
    for (item, support) in frequentItems:
        prefix = [[item]]
        result[toStr(prefix)] = support
        result.update( prefixSpanRecursion(prefix,projectDatabase(sequence_DB,prefix), min_sup) )
    return result

""" 
    recursive call of prefixspan
    input: projected sequence db, minimum support, current prefix
"""
def prefixSpanRecursion(prefix,current_DB, min_sup):
    result = {}
    #print('call of recursion with prefix: ', prefix, 'and db: ', current_DB)
    
    # "assemble to last element of prefix": add to last timestep (happens at the same time)
    frequentItemsSameEvent = getFrequentItems(current_DB,min_sup,prefix[-1], False)
    #print('frequent item same event: ', frequentItemsSameEvent)
    for (item, support) in frequentItemsSameEvent:
        if all(item>x for x in prefix[-1]):
            #print('item: ', item, 'support: ', support)
            #print('prefix before: ', prefix)
            newPrefix = copy.deepcopy(prefix)
            newPrefix[-1].append(item)
            #print('new prefix:', newPrefix, 'count:',support )
            result[toStr(newPrefix)] = support
            result.update(prefixSpanRecursion(newPrefix, projectDatabase(current_DB,[newPrefix[-1]], False),min_sup))

    # "append to prefix": add new timestep
    frequentItemsNewEvent = getFrequentItems(current_DB,min_sup,prefix[-1], True)
    #print('frequentItemsNewEvent: ', frequentItemsNewEvent)
    for (item, support) in frequentItemsNewEvent: 
        newPrefixB = copy.deepcopy(prefix)
        newPrefixB.append([item])
        #print('new prefix:', newPrefixB, 'count:',support )
        result[toStr(newPrefixB)] = support
        result.update(prefixSpanRecursion(newPrefixB, projectDatabase(current_DB,[newPrefixB[-1]], True),min_sup))
         
    return result

In [114]:
datasetValidate = [
    [["a"], ["a", "b", "c"], ["a", "c"], ["c"]],
    [["a"], ["c"], ["b", "c"]],
    [["a", "b"], ["d"], ["c"], ["b"], ["c"]],
    [["a"], ["c"], ["b"], ["c"]]
]

In [115]:
result = prefixSpan(datasetValidate, 2)

In [116]:
result # matches results from algorithm found here: http://sequenceanalysis.github.io/

{'[[a,b],[c],[c]]': 2,
 '[[a,b],[c]]': 2,
 '[[a,b]]': 2,
 '[[a],[b,c]]': 2,
 '[[a],[b],[c]]': 3,
 '[[a],[b]]': 4,
 '[[a],[c],[b],[c]]': 2,
 '[[a],[c],[b]]': 3,
 '[[a],[c],[c]]': 4,
 '[[a],[c]]': 4,
 '[[a]]': 4,
 '[[b,c]]': 2,
 '[[b],[c],[c]]': 2,
 '[[b],[c]]': 3,
 '[[b]]': 4,
 '[[c],[b],[c]]': 2,
 '[[c],[b]]': 3,
 '[[c],[c]]': 4,
 '[[c]]': 4}

In [103]:
len(result)

19